Sources/SPMTestSupport/MockWorkspace.swift

```diff
@@ -115,7 +118,17 @@ public final class MockWorkspace {
             case .fileSystem(let path):
                 packagePath = basePath.appending(path)
             case .sourceControl(let url):
-                packagePath = basePath.appending(components: "sourceControl", url.absoluteString.spm_mangledToC99ExtendedIdentifier())
+                if let containerProvider = customPackageContainerProvider {
+                    let observability = ObservabilitySystem.makeForTesting()
+                    let packageRef = PackageReference(identity: PackageIdentity(url: url), kind: .remoteSourceControl(url))
+                    let container = try temp_await { containerProvider.getContainer(for: packageRef, skipUpdate: true, observabilityScope: observability.topScope, on: .sharedConcurrent, completion: $0) }
+                    guard let customContainer = container as? CustomPackageContainer else {
+                        throw StringError("invalid custom container: \(container)")
+                    }
+                    packagePath = try customContainer.retrieve(at: try Version(versionString: package.versions.first!!), observabilityScope: observability.topScope)
+                } else {
+                    packagePath = basePath.appending(components: "sourceControl", url.absoluteString.spm_mangledToC99ExtendedIdentifier())
+                }
```

Comment:
maybe a FIXME or a radar to clean this up a bit?

Reply:
rdar://86349395